src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/X509AsymmetricSecurityKey.cs

```diff
@@ -0,0 +1,450 @@
+                            // have access to DSACertificateExtensions 
+                        }
+
+                        if (_certificate.HasPrivateKey && _privateKey == null)
+                            DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.PrivateKeyNotSupported));
+                        _privateKeyAvailabilityDetermined = true;
+                    }
+                }
+                return _privateKey;
+            }
+        }
+
+        private AsymmetricAlgorithm PublicKey
+        {
+            get
+            {
+                if (!_publicKeyAvailabilityDetermined)
+                {
+                    lock (ThisLock)
+                    {
+                        if (!_publicKeyAvailabilityDetermined)
+                        {
+                            _publicKey = _certificate.GetRSAPublicKey();
+                            if (_publicKey == null)
+                            {
+                                // Need DSACertificateExtensions  to support DSA certificate which is in netstandard2.1 and netcore2.0. As we target netstandard2.0, we don't
+                                // have access to DSACertificateExtensions 
+                                DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.PublicKeyNotSupported));
+                            }
+
+                            _publicKeyAvailabilityDetermined = true;
+                        }
+                    }
+                }
+                return _publicKey;
+            }
+        }
+
+        private Object ThisLock
+        {
+            get
+            {
+                return _thisLock;
+            }
+        }
+
+        public override byte[] DecryptKey(string algorithm, byte[] keyData)
+        {
+            throw ExceptionHelper.PlatformNotSupported();
+        }
+
+        public override byte[] EncryptKey(string algorithm, byte[] keyData)
+        {
+            throw ExceptionHelper.PlatformNotSupported();
+        }
+
+        public override AsymmetricAlgorithm GetAsymmetricAlgorithm(string algorithm, bool privateKey)
+        {
+            if (privateKey)
+            {
+                if (this.PrivateKey == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.MissingPrivateKey));
+                }
+
+                if (string.IsNullOrEmpty(algorithm))
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgument(algorithm, SR.Format(SR.EmptyOrNullArgumentString, nameof(algorithm)));
+                }
+
+                switch (algorithm)
+                {
+                    case SignedXml.XmlDsigDSAUrl:
+                        if ((this.PrivateKey as DSA) != null)
+                        {
+                            return (this.PrivateKey as DSA);
```

Comment:
Save the value of the `as` to return it here?  `if ((this.PrivateKey as DSA dsa) != null) { return dsa; }` ?